utils/src/protobuf/pmgdMessages.proto

```diff
@@ -207,6 +207,11 @@ message LinkInfo
     // are unique. Does carry a performance penalty to do that
     // in cases where it is known that neighbors are unique.
     bool nb_unique = 14;
+
+    // Specify edge constraints
+    // Specify search conditions and whether to OR or AND them
+    PredicateOp p_op = 20;
+    repeated PropertyPredicate predicates = 21;
```

Comment:
This branch overall makes me concerned about breaking the logic we had behind our API. We had tried to define link and other blocks to be common pieces to build commands. With this predicate op introduction here, it is an extra thing when you do AddEntity with a link, predicate op makes no sense there. properties would make sense in that scenario. But in order to avoid such specializations we had said, to do more complex things, use FindConnection and the link block was meant to stay for simple scenarios when your edges would not have too many things.

Reply:
I would have tried to achieve the goal you are going after, by making sure that FindEntity could understand the endpoints of FindConnection and FindConnection iterator had a way to first look at srcni and then predicates, instead of putting constraint on link block directly. IT probably would have amounted to fewer changes overall and not introduced nesting in the blocks.

Reply:
Since FindConnection already applies property filter, query_node() would need to add a condition to check if ref belongs in _cached_edges() and basically do a get_source() or get_destination() with edge traversal in there. So a new build_results with edge iterator looking at source/destination node.

Reply:
We can discuss this on the phone if you want.  Perhaps I am missing something in my understanding as to why this way.

Reply:
You're right about the alternatives. Luis and I have discussed them and we should have that same discussion again with you.